src/transcripts.rs

```diff
@@ -1,76 +1,199 @@
+/// The prover uses this whenever it needs a random nonce.
+/// The batch verifier uses this to generate weights.
+pub(crate) struct RangeProofTranscript<P>
+where
+    P: Compressable + Precomputable,
+    P::Compressed: FixedBytesRepr + IsIdentity,
+{
+    transcript: Transcript,
+    bytes: Option<Zeroizing<Vec<u8>>>,
+    _phantom: PhantomData<P>,
+}
+
+impl<P> RangeProofTranscript<P>
 where
     P: Compressable + Precomputable,
     P::Compressed: FixedBytesRepr + IsIdentity,
 {
-    transcript.validate_and_append_point(b"H", h_base_compressed)?;
-    for item in g_base_compressed {
-        transcript.validate_and_append_point(b"G", item)?;
+    /// Initialize a transcript.
+    ///
+    /// The prover should include its `witness` here; the verifier should pass `None`.
+    #[allow(clippy::too_many_arguments)]
+    pub(crate) fn new<R: CryptoRngCore>(
+        label: &'static str,
+        h_base_compressed: &P::Compressed,
+        g_base_compressed: &[P::Compressed],
+        bit_length: usize,
+        extension_degree: usize,
+        aggregation_factor: usize,
+        statement: &RangeStatement<P>,
+        witness: Option<&RangeWitness>,
+        external_rng: &mut R,
+    ) -> Result<(Self, TranscriptRng), ProofError> {
+        // Initialize the transcript with parameters and statement
+        let mut transcript = Transcript::new(label.as_bytes());
+        transcript.domain_separator(b"Bulletproofs+", b"Range Proof");
+        transcript.validate_and_append_point(b"H", h_base_compressed)?;
+        for item in g_base_compressed {
+            transcript.validate_and_append_point(b"G", item)?;
+        }
+        transcript.append_u64(b"N", bit_length as u64);
+        transcript.append_u64(b"T", extension_degree as u64);
+        transcript.append_u64(b"M", aggregation_factor as u64);
+        for item in &statement.commitments_compressed {
+            transcript.append_point(b"Ci", item);
+        }
+        for item in &statement.minimum_value_promises {
+            if let Some(minimum_value) = item {
+                transcript.append_u64(b"vi - minimum_value", *minimum_value);
+            } else {
+                transcript.append_u64(b"vi - minimum_value", 0);
+            }
+        }
+
+        // Serialize the witness if provided
+        let bytes = if let Some(witness) = witness {
+            let size: usize = witness
+                .openings
+                .iter()
+                .map(|o| size_of::<u64>() + o.r.len() * size_of::<Scalar>())
+                .sum();
+            let mut witness_bytes = Zeroizing::new(Vec::<u8>::with_capacity(size));
+            for opening in &witness.openings {
+                witness_bytes.extend(opening.v.to_le_bytes());
+                for r in &opening.r {
+                    witness_bytes.extend(r.as_bytes());
+                }
+            }
+
+            Some(witness_bytes)
+        } else {
+            None
+        };
```

Comment:
nit: (optional) 

```rust
impl RangeWitness {

pub(super) fn to_non_canonical_bytes(&self) -> Zeroize<Vec<u8>> {
    let size: usize = witness
                .openings
                .iter()
                .map(|o| size_of::<u64>() + o.r.len() * size_of::<Scalar>())
                .sum();
            let mut witness_bytes = Zeroizing::new(Vec::<u8>::with_capacity(size));
            for opening in &witness.openings {
                witness_bytes.extend(opening.v.to_le_bytes());
                for r in &opening.r {
                    witness_bytes.extend(r.as_bytes());
                }
            }

            witness_bytes
}
}
```

```suggestion
        let witness_bytes = witness.map(|w| w.to_non_canonical_bytes())
```

Reply:
That's more or less what I had in an initial version. I took the byte serialization out of `RangeWitness` since it's not suitable (and possibly not safe) to use it in other contexts. Because it's so tightly linked to the transcript use case, it seemed safer to lock it away in `RangeProofTranscript` instead.

Interestingly, we could just make the serialization canonical if we really wanted to; it wouldn't matter from a transcript perspective. But that seems unnecessary since serializing secret data isn't something that should be done without a good reason.